### ISSUE TITLE
Investigate and fix test failure when bumping to JupyterHub 4.0.0b1

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#3046]
JupyterHub 4.0.0b1 is out with the [changelog here](https://jupyterhub.readthedocs.io/en/latest/reference/changelog.html). It seems though that we have a test failure and I couldn't quickly figure out what it related to.

- https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/3045

The failure surfaces as being stuck at repeatedly printing `Awaiting ...` below:

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/29b041ced53678357d062c188f52e8983bfb1f03/tests/test_spawn.py#L196-L215

A good steps to resolving this is probably to try capture logs of a failed startup of the user pod. Currently this isn't captured currently as its removed before the `Kubernetes namespace report` step is run at the end of the failed github workflow's job. So, in practice I think a local development setup is relevant.

**Comment by consideRatio** [comment]
### Logs from singleuser server startup failing

```
[I 2023-03-02 12:50:54.158 SingleUserLabApp mixins:547] Starting jupyterhub single-user server version 4.0.0b1
[I 2023-03-02 12:50:54.158 SingleUserLabApp mixins:561] Extending jupyterlab.labhubapp.SingleUserLabApp from jupyterlab 3.4.8
[I 2023-03-02 12:50:54.158 SingleUserLabApp mixins:561] Extending jupyter_server.serverapp.ServerApp from jupyter_server 1.23.6
[D 2023-03-02 12:50:54.162 SingleUserLabApp application:190] Searching ['/home/jovyan/.jupyter', '/home/jovyan/.local/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 2023-03-02 12:50:54.162 SingleUserLabApp application:902] Looking for jupyter_config in /etc/jupyter
[D 2023-03-02 12:50:54.162 SingleUserLabApp application:902] Looking for jupyter_config in /usr/local/etc/jupyter
[D 2023-03-02 12:50:54.162 SingleUserLabApp application:902] Looking for jupyter_config in /home/jovyan/.local/etc/jupyter
[D 2023-03-02 12:50:54.162 SingleUserLabApp application:902] Looking for jupyter_config in /home/jovyan/.jupyter
[D 2023-03-02 12:50:54.163 SingleUserLabApp application:902] Looking for jupyter_server_config in /etc/jupyter
[D 2023-03-02 12:50:54.163 SingleUserLabApp application:902] Looking for jupyter_server_config in /usr/local/etc/jupyter
[D 2023-03-02 12:50:54.163 SingleUserLabApp application:902] Looking for jupyter_server_config in /home/jovyan/.local/etc/jupyter
[D 2023-03-02 12:50:54.163 SingleUserLabApp application:902] Looking for jupyter_server_config in /home/jovyan/.jupyter
[D 2023-03-02 12:50:54.164 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_server_config: 
    	/etc/jupyter/jupyter_server_config.json
[D 2023-03-02 12:50:54.164 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_server_config: 
    	/usr/local/etc/jupyter/jupyter_server_config.d/jupyterlab.json
    	/usr/local/etc/jupyter/jupyter_server_config.d/nbclassic.json
    	/usr/local/etc/jupyter/jupyter_server_config.d/nbgitpuller.json
    	/usr/local/etc/jupyter/jupyter_server_config.d/notebook_shim.json
    	/usr/local/etc/jupyter/jupyter_server_config.d/retrolab.json
    	/usr/local/etc/jupyter/jupyter_server_config.json
[D 2023-03-02 12:50:54.164 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_server_config: 
    	/home/jovyan/.local/etc/jupyter/jupyter_server_config.json
[D 2023-03-02 12:50:54.164 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_server_config: 
    	/home/jovyan/.jupyter/jupyter_server_config.json
[I 2023-03-02 12:50:54.171 SingleUserLabApp manager:344] jupyterlab | extension was successfully linked.
[I 2023-03-02 12:50:54.172 SingleUserLabApp manager:344] nbclassic | extension was successfully linked.
[I 2023-03-02 12:50:54.173 SingleUserLabApp manager:344] nbgitpuller | extension was successfully linked.
[D 2023-03-02 12:50:54.366 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_notebook_config: 
    	/home/jovyan/.jupyter/jupyter_notebook_config.json
[D 2023-03-02 12:50:54.366 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
[D 2023-03-02 12:50:54.366 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/nbgitpuller.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/retrolab.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
[D 2023-03-02 12:50:54.366 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_notebook_config: 
    	/home/jovyan/.local/etc/jupyter/jupyter_notebook_config.json
[D 2023-03-02 12:50:54.366 SingleUserLabApp config_manager:93] Paths used for configuration of jupyter_notebook_config: 
    	/home/jovyan/.jupyter/jupyter_notebook_config.json
[I 2023-03-02 12:50:54.367 SingleUserLabApp manager:344] notebook_shim | extension was successfully linked.
[I 2023-03-02 12:50:54.370 SingleUserLabApp manager:344] retrolab | extension was successfully linked.
[I 2023-03-02 12:50:54.379 SingleUserLabApp manager:362] notebook_shim | extension was successfully loaded.
[I 2023-03-02 12:50:54.379 LabApp] JupyterLab extension loaded from /usr/local/lib/python3.11/site-packages/jupyterlab
[I 2023-03-02 12:50:54.379 LabApp] JupyterLab application directory is /usr/local/share/jupyter/lab
[I 2023-03-02 12:50:54.380 SingleUserLabApp manager:362] jupyterlab | extension was successfully loaded.
[I 2023-03-02 12:50:54.382 SingleUserLabApp manager:362] nbclassic | extension was successfully loaded.
[I 2023-03-02 12:50:54.382 SingleUserLabApp manager:362] nbgitpuller | extension was successfully loaded.
[I 2023-03-02 12:50:54.383 SingleUserLabApp manager:362] retrolab | extension was successfully loaded.
Traceback (most recent call last):
  File "/usr/local/bin/jupyterhub-singleuser", line 8, in <module>
    sys.exit(main())

  _   _          _      _
 | | | |_ __  __| |__ _| |_ ___
 | |_| | '_ \/ _` / _` |  _/ -_)
  \___/| .__/\__,_\__,_|\__\___|
       |_|
                                                                           
Read the migration plan to Notebook 7 to learn about the new features and the actions to take if you are using extensions.

https://jupyter-notebook.readthedocs.io/en/latest/migrate_to_notebook7.html

Please note that updating to Notebook 7 might break some of your extensions.

             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/jupyterhub/singleuser/app.py", line 76, in main
    return SingleUserLabApp.launch_instance()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/jupyter_core/application.py", line 277, in launch_instance
    return super().launch_instance(argv=argv, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/traitlets/config/application.py", line 1042, in launch_instance
    app.initialize(argv)
  File "/usr/local/lib/python3.11/site-packages/jupyterhub/singleuser/mixins.py", line 951, in initialize
    result = super().initialize(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/jupyterhub/singleuser/mixins.py", line 608, in initialize
    self.patch_templates()
  File "/usr/local/lib/python3.11/site-packages/jupyterhub/singleuser/mixins.py", line 755, in patch_templates
    with template_dir.joinpath("page.html").open() as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/pathlib.py", line 1044, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.11/site-packages/jupyterhub/singleuser/templates/page.html'
```

**Comment by consideRatio** [comment]
I reported this failure, which stems from jupyterhub 4 + jupyter_server 1 in the base image in https://github.com/jupyterhub/jupyterhub/issues/4384.

**Comment by consideRatio** [comment]
Resolved upstream now and fixed in 4.0.0b2
